Fix size_hint call sites failing on unbacked SymInts

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110520

Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler

aakhundov committed:
Add more size_hint fallbacks for unbacked SymInts

Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

aakhundov committed:
Update on "Add more size_hint fallbacks for unbacked SymInts"


Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

eellison reviewed (commented):
We should be constraining the hint with the value range if it exists, cc @ezyang.

ezyang commented on torch/_inductor/dependencies.py:
```diff
@@ -49,7 +49,7 @@ def numbytes_hint(self):
                 if var in vars:
                     numel = numel * size
-        return V.graph.sizevars.size_hint(numel) * get_dtype_size(
+        return V.graph.sizevars.size_hint(numel, fallback=8192) * get_dtype_size(
```
I don't think this is sound, see

```
    def can_inplace(self, read_dep: dependencies.MemoryDep):
        if self.get_aliases() or self.is_template():
            return False
                
        if read_dep.name not in self.scheduler.name_to_node:
            # don't allow reuse of an 'input' buffer, we don't own it
            # (would this have been fixed if I tracked mutations properly above?)
            return False
            
        if not isinstance(
            self.node, (torch._inductor.ir.AllReduce, torch._inductor.ir.InPlaceHint)
        ):
            # TODO make this a property of the IR
            return False

        if len(self.read_writes.writes) == 1:
            write_dep = next(iter(self.read_writes.writes))
            return read_dep.numbytes_hint() == write_dep.numbytes_hint()

        return False
```

We must know for sure that the sizes are the same to inplace, so we shouldn't fallback, we should reject inplacing (or maybe do it symbolically?)

aakhundov replied:
Is the problem you're referring to specifically related to `can_inplace` (in which case we can, indeed, reject inplacing in `can_inplace` in the face of unbacked SymInts)? Or are there other cases where using the `fallback=8192` inside `numbytes_hint` can cause problems (in which case we can make `fallback=None` a parameter of `numbytes_hint`, propagate it to `size_hint`, and set it to `8192` only from unproblematic places)?

ezyang replied:
`can_inplace` uses `numbytes_hint` which is where you made the fallback use. So you  just need to fix `can_inplace` to use something else. But it looks like this function is actually the sole user.

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -1769,7 +1769,7 @@ def _dynamic_reshape_indexer(old_size, new_size):
         Perform a reshape entirely by modifying indexing math
         """
-        size_hint = V.graph.sizevars.size_hint
+        size_hint = partial(V.graph.sizevars.size_hint, fallback=8192)
```
This looks pretty sus. Can we just disable dynamic reshape indexing when things are unbacked?

aakhundov replied:
Could you please clarify what "disable" mean in this context? I.e., I see `dynamic_reshape_indexer` being used in a few places in the IR like this:

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py?#L996

or here

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py?#L1705

and there doesn't seem to be obvious ways to get around those calls. Or should we invoke some other (static) reshape indexer within `dynamic_reshape_indexer` if we see unbacked SymInts?

ezyang replied:
Hmm, I am not too sure. @jansel any ideas?

jansel replied:
The dynamic_reshape_indexer is used to implement views.

I think we could just call `.realize()` as a fallback if there are unbacked symints.

ezyang reviewed (changes requested):
The latter two changes look OK, but the first two seem trouble.

t10-13rocket commented:
Ms.AmyV.


On Wed, Oct 4, 2023 at 11:34 AM Adnan Akhundov ***@***.***>
wrote:

> Stack from ghstack <https://github.com/ezyang/ghstack> (oldest at bottom):
>
>    - *->* #110520 <https://github.com/pytorch/pytorch/pull/110520>
>
> Summary: Unbacked SymInts can't get a sizevars.size_hint due to being
> data-dependent. #109893 <https://github.com/pytorch/pytorch/pull/109893>
> has added a new fallback parameter to sizevars.size_hint to specify the
> fallback value in cases like unbacked SymInt. In this PR we add more of
> those.
>
> Test Plan: CI
>
> Reviewers:
>
> Subscribers:
>
> Tasks:
>
> Tags:
>
> cc @voznesenskym <https://github.com/voznesenskym> @penguinwu
> <https://github.com/penguinwu> @EikanWang <https://github.com/EikanWang>
> @jgong5 <https://github.com/jgong5> @Guobing-Chen
> <https://github.com/Guobing-Chen> @XiaobingSuper
> <https://github.com/XiaobingSuper> @zhuhaozhe
> <https://github.com/zhuhaozhe> @blzheng <https://github.com/blzheng>
> @Xia-Weiwen <https://github.com/Xia-Weiwen> @wenzhe-nrv
> <https://github.com/wenzhe-nrv> @jiayisunx <https://github.com/jiayisunx>
> @peterbell10 <https://github.com/peterbell10> @ipiszy
> <https://github.com/ipiszy> @yf225 <https://github.com/yf225> @chenyang78
> <https://github.com/chenyang78> @kadeng <https://github.com/kadeng>
> @muchulee8 <https://github.com/muchulee8> @ColinPeppler
> <https://github.com/ColinPeppler>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/pytorch/pytorch/pull/110520
> Commit Summary
>
>    - 0f6aba8
>    <https://github.com/pytorch/pytorch/pull/110520/commits/0f6aba8bd03fb170a585837029c073d6b638b891>
>    Add more size_hint fallbacks for unbacked SymInts
>
> File Changes
>
> (4 files <https://github.com/pytorch/pytorch/pull/110520/files>)
>
>    - *M* torch/_inductor/dependencies.py
>    <https://github.com/pytorch/pytorch/pull/110520/files#diff-eac5a554f9c6b61ad0991e1027bddeb09204100f4790021e2e653f9a59dd8096>
>    (4)
>    - *M* torch/_inductor/ir.py
>    <https://github.com/pytorch/pytorch/pull/110520/files#diff-cf6ca00beddc32a2a6a2933fb9913b6a2b925ffc3b745488967210e4343134ac>
>    (2)
>    - *M* torch/_inductor/kernel/mm_common.py
>    <https://github.com/pytorch/pytorch/pull/110520/files#diff-8973a6537701d24baabc55ae4c64093f4762a30e4305fddbbbba9d4b54fca3a1>
>    (6)
>    - *M* torch/_inductor/lowering.py
>    <https://github.com/pytorch/pytorch/pull/110520/files#diff-a1b077971cddfabfa0071c5162265066e867bc07721816d95b9cbe58431c38e3>
>    (2)
>
> Patch Links:
>
>    - https://github.com/pytorch/pytorch/pull/110520.patch
>    - https://github.com/pytorch/pytorch/pull/110520.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/110520>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARC2T26HWMTISYNC32W7NFLX5V6Z3AVCNFSM6AAAAAA5S2J26OVHI2DSMVQWIX3LMV43ASLTON2WKOZRHEZDMNBWGE4TEMA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

aakhundov commented:
> We should be constraining the hint with the value range if it exists, cc @ezyang.

@eellison do you think this is universal enough to put it into size_hits (e.g., by clipping the fallback by min / max if those exist)?

eellison commented:
@aakhundov yea you could clip it within the size hint api

aakhundov committed:
Update on "Add more size_hint fallbacks for unbacked SymInts"


Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov committed:
Update on "Add more size_hint fallbacks for unbacked SymInts"


Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
@ezyang @eellison Updates since the last time:

1. Clip the `fallback` (int) value in the `size_hint(expr, fallback)`  by the value range for the `expr` if available.

2. Ditch the `fallback` in `numbytes_hint` and fix the call sites in the `scheduler.py` instead.

3. Ditch the `fallback` in `_dynamic_reshape_indexer` and realize the input if there are unbacked SymInts in input or output size of the view op.

Please take a look. Thanks!

aakhundov committed:
Update on "Fix size_hint call sites failing on unbacked SymInts"


Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -1717,14 +1717,25 @@ def create(cls, x, new_size):
+            if unbacked_symbols_in_sizes and (not is_contiguous_storage_and_layout(x)):
+                # realize x; otherwise, the dynamic_reshape_indexer below will fail
+                # due to the size_hint's inability to process unbacked SymInts
+                x.realize()
```
@Chillee and @aakhundov discussed this in chat, the thinking is that realize should be initially contiguous as long as no one makes it noncontiguous later.

ezyang commented on torch/_inductor/scheduler.py:
```diff
@@ -583,6 +582,12 @@ def can_inplace(self, read_dep: dependencies.MemoryDep):
+                return False
+            if write_dep.has_unbacked_symbols():
+                # can't inplace, as can't guarantee the exact size match
+                return False
```
We can probably make this a little better by doing an exact symbolic match, but this is fine for now.

aakhundov replied:
Actually, if we take symbolic product of the `read_dep` and `write_dep` shapes with unbacked SymInts in them, and those two symbolic products turn out to be equal, this means that the buffer sizes are provably the same and we can allow in-placing? I this is the case, I think, it's not too hard to do.

On the other hand, I have doubts about the current model for backed SymInts. We're relying on the size hints of the numels, but size hints being the same doesn't necessarily mean that the shapes are the same? I guess, matching symbolic numels would be more reliable?

ezyang replied:
> Actually, if we take symbolic product of the read_dep and write_dep shapes with unbacked SymInts in them, and those two symbolic products turn out to be equal, this means that the buffer sizes are provably the same and we can allow in-placing? I this is the case, I think, it's not too hard to do.

Exactly.

> On the other hand, I have doubts about the current model for backed SymInts. We're relying on the size hints of the numels, but size hints being the same doesn't necessarily mean that the shapes are the same? I guess, matching symbolic numels would be more reliable?

Yeah, this seems a bit sus. Suppose that I have s0 and s1, and they happen to be the same this run. Why would reusing the buffer be OK if s0 and s1 later vary? Maybe we can construct a test case that forces a failure here...

ezyang commented on torch/_inductor/scheduler.py:
```diff
@@ -1624,6 +1629,9 @@ def score_fusion_memory(self, node1, node2):
             node2.read_writes.reads | node2.read_writes.writes
         )
+        common_memory_deps = {
+            dep for dep in common_memory_deps if not dep.has_unbacked_symbols()
```
For this one, wouldn't you prefer to use a fallback heuristic number?

aakhundov replied:
I'm somewhat hesitant to use a heuristic number here: if the number is too large, this may lead to overoptimistic saving estimation; if too small, this likely won't make a substantial difference. So my thinking was to (somewhat conservatively) drop the dependencies with unbacked SymInts in them, as no reasonably close estimation can be done (like with backed SymInts). Does this make sense?

ezyang replied:
ok

ezyang commented on torch/_inductor/lowering.py:
```diff
@@ -748,16 +748,19 @@ def expand(x, sizes):
+        # this cannot be done directly as below as we'll choke on the size_hint
+        # here
+        if x_size_product > 0 and not any(
+            V.graph.sizevars.shape_env.is_unbacked_symint(s) for s in sizes
```
Is this test redundant now?

aakhundov replied:
I think, `x.get_size()` and `sizes` can generally be different in the `expand` op, with unbacked SymInts being present in any of them, but not the other? That's why I kept both tests.

ezyang replied:
you are right

ezyang commented:
There are some suboptimal bits, but nothing that looks wrong. Thanks for doing this.

It would be really good to have tests for this, because this is from some internal model and we really don't want our only test coverage for this to be when we actually run it for real.

aakhundov committed:
Update on "Fix size_hint call sites failing on unbacked SymInts"


Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov committed:
Update on "Fix size_hint call sites failing on unbacked SymInts"


Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov committed:
Update on "Fix size_hint call sites failing on unbacked SymInts"


Summary: Unbacked SymInts can't get a `sizevars.size_hint` due to being data-dependent. #109893 has added a new `fallback` parameter to `sizevars.size_hint` to specify the fallback value in cases like unbacked SymInt. In this PR we add more of those.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge
